Add NAT traversal spike (#60)
comit-network/spikes: 0025-nat-traversal.adoc

== Goals

1. Allow a comit node to traverse a NAT to reach another comit node.

== Abbreviations

NAT: Network address translation

UPnP: Universal plug and play

TURN: Traveral using relays around NAT

STUN: Session traversal utilities around NAT

ICE: Interactive connectivity Establishment

== NAT Traversal Methods

This sections provides an overview of current NAT traversal methods. Alice is the dialer and Bob is the receiver behind a NAT device.

=== Manual Port Forwarding

Bob configures his router to forward an externally facing port to his device.
Bob notifies peers of his public IP and forwarded port so that they can dial him.
This solution does not require any changes to the COMIT codebase and places the burden on the user to configure their router.
Users who cannot access the router, for example users on a corporate or mobile network will not be able to use this method.

- - Requires the user to have access to NAT configuration

- - Requires the user to configure NAT settings

- + Alice and Bob can directly communicate through a direct connection

=== Automatic Port Forwarding using UPnP

> There is no single perfect solution to solving the NAT traversal problem.There is no single perfect solution to solving the NAT traversal problem.

Reviewer: Something is odd here :D 
```suggestion
There is no single perfect solution to solving the NAT traversal problem.
```

If both Alice and Bob are behind a NAT a third peer which is not behind NAT can be used as a relay.
Alice and Bob both dial a relay peer. Th
This technique is used by the Bittorrent network and relies on there being a critical mass of relay peers that have a public IP.

- + Does not require the use to have configure to the NAT settings on their router

- + Does not require the user to have access to NAT configuration

- - Does not require the user to have access to NAT configuration

- - Requires Bob to have a known external IP

=== STUN/Hole Punching

Bob sends UDP messages to a third peer(STUN server).
The STUN server will return which ports and public IP Bobs's UDP messages originated from.
Bob has discovered public IP and ports which peers can use to dial him.
Bob

STUN cannot be used to traverse NAT configurations that randomnly assign port and/or IP for new connections.
A symmetric NAT will randomnly generate a new port and/or public IP for each outbound connection made by Alice.
This prevents Bob from determining the external IP and ports assigned to her by her NAT.
Symmetric NAT is commonly used by mobile network providers.

> - + Does not require the user to configure NAT settiThere is no single perfect solution to solving the NAT traversal problem.ngs

Reviewer: ```suggestion
- + Does not require the user to configure NAT settings. There is no single perfect solution to solving the NAT traversal problem.
```

Author: woops

- + Alice and Bob can directly communicate through a direct connection

- - Requires a third peer to facilitate a connection between two peers

- - Cannot traverse symmetric, cone and restricted cone type NATs

- - Requires a third peer to start a connection between Alice and Bob

- - Requires the third peer to know its inbound ports (either through another STUN/TURN or port forwarding)

=== TURN

If STUN cannot be used to traverse a NAT, a relay peer can be used a relay between the comit_nodes.
When Bob isn’t able to listen on a public address, Bob can dial out to a relay peer, which will keep a long-lived connection open.
Other peers will be able to dial through the relay peer, which will forward traffic to Bob.

- + Can be used to traverse all commonly know NAT configurations

- + Does not require the use to have configure NAT settings

- - Requires a third peer to serve as a relay for communications between Alice and Bob

- - Requires the third peer to know its inbound ports (either through another STUN/TURN or port forwarding)

> == Recommendation

Reviewer 2: This sounds more like a summary to me :)

Manual and automatic port forwarding solve the NAT traversal problem through hardware configuration.
We cannot rely on all users to have access to their network hardware and therefore we will need to implement alternative NAT traversal techniques.

Reverse connections cannot be used because many of our protocols require both peers to be able to initate a connection with each other.
This means that COMIT protocols will have to redesigned to use the connection initiated by dialer to complete a SWAP.
Peers behind a NAT device will be unable to play an active role in the COMIT network.

In order to allow comit nodes to traverse any NAT configuration the TURN protocol will need to be implemented.
ICE recommends that peers attempt to connect using STUN if possible only use TURN a connection using STUN cannot be established.
STUN is preferred over TURN for performance and efficiency reasons as it does require a third party to relay communications between nodes.

Although TURN servers can resolve the NAT traversal problem, at least one STUN or TURN server that knows its own external ports and IP will be required to bootstrap the network and to restart the network in case it goes down.

> Users who have knowledge of their public IP and ports should have the option to run as STUN and/or TURN servers.

Reviewer: I'm not sure if I follow :) 

Do you propose that we provide both solutions (STUN and TURN)? 

Can you rephrase this section so that it is clear what your recommendation is please? Same goes for the next section.

Author: yes we should ideally provide both solutions if want to be able to traverse most known NAT configurations. I'll reword this section to make it clear

= NAT Traversal
Rishab Sharma<rishab.sharma@tenx.tech>;
:toc:
:revdate: 2020-03-15

NOTE: Author: {authors} +
Date: {revdate} +
Tracking issue: https://github.com/comit-network/comit-rs/issues/1172[#1172]

== Abbreviations

> NAT: Network address translation/translator

Reviewer 3: ```suggestion
NAT: Network Address Translation/Translator
```

> UPnP: Universal plug and play

Reviewer 3: ```suggestion
UPnP: Universal Plug and Play
```

> TURN: Traveral using relays around NAT

Reviewer 3: ```suggestion
TURN: Traversal Using Relays around NAT
```

> STUN: Session traversal utilities around NAT

Reviewer 3: ```suggestion
STUN: Session Traversal Utilities for NAT
```

This solution does not require any changes to the COMIT codebase and places the burden on the user to configure their NAT.
Users who do not have control over the configuration of their NAT can not use this method.
We cannot completely rely on port forwarding because users on mobile, shared or corporate managed networks will not have access to NAT configuration.

- + Alice and Bob can communicate through a direct connection

- - Requires Bob to have access to NAT configuration

- - Requires Bob to configure his NAT

=== Automatic Port Forwarding using UPnP

A UPnP enabled NAT automatically forwards publicly facing ports as requested by an application.
Unlike manual port forwarding, Bob is not required to configure his NAT.
Not all NATs are UPnP capable and it is often disabled or not used by peer to peer applications.
Bitcoin core removed UPnP support due to security concerns: https://bitcoin.org/en/alert/2015-10-12-upnp-vulnerability.

- + Does not require Bob to configure his NAT

- + Alice and Bob can communicate through a direct connection

- - Requires the user to have a UPnP capable and enabled NAT (UPnP capable routers are fairly common in home networks)

> - - Opens up the user to security vulnerabilities arising from the NAT blindly trusting applications and allowing to open externally facing ports

Reviewer 3: ```suggestion
- - Opens up the user to security vulnerabilities arising from the NAT blindly trusting applications and allowing incoming traffic to connect to externally facing ports
```

=== Reverse connection

> Instead of Alice attempting to reach Bob through his NAT. Bob dials Alice and Alice uses the connection to send to data to Bob.

Reviewer 3: ```suggestion
Instead of Alice attempting to reach Bob through his NAT. Bob dials Alice and Alice uses the connection to send data to Bob.
```

> This technique only works if Alice knows her public IP and port and has provided it to Bo.

Reviewer 3: ```suggestion
This technique only works if Alice knows her public IP and port and has provided it to Bob.
```

> Since Alice is unable to dial Bob, Alice is unable to initiate communication and must wait to be dialled by a Bob.

Reviewer 3: ```suggestion
Since Alice is unable to dial Bob, Alice is unable to initiate communication and must wait to be dialled by Bob.
```

> Reverse connections cannot be used because many of our protocols require both peers to be able to initate a connection with each other.

Reviewer 2: That is not quite true.

Libp2p is specifically designed so you can open substreams in either direction without having to care about the physical connection.

If for some reason we already have a physical connection to a node, then any of the peers can use this with any kind of protocol on top.

This means that COMIT protocols will have to redesigned to use the connection initiated by dialer to complete an atomic swap.
Peers behind a NAT will be unable to play an active role in the COMIT network.

- + Does not require Bob to configure his NAT

- + Does not require the user to have access to NAT configuration

- + Alice and Bob can  communicate through a direct connection

- - Alice can not dial Bob and therefore cannot initiate a swap or use Bob as a peer to propagate information to the network

- - We will most likely need to redesign protocols to utilise reverse connections

- - Requires Bob to have a known external IP

=== STUN/Hole Punching

> Bob sends UDP messages to a third peer known as a STUN server.

Reviewer 3: ```suggestion
Bob sends UDP messages to a third party known as a STUN server.
```

> The STUN server will return which ports and public IP Bobs's UDP messages originated from.

Reviewer 3: ```suggestion
The STUN server will return which port and IP Bobs's UDP message originated from.
```

> Bob has now discovered public IP and ports which peers can use to dial him.

Reviewer 3: ```suggestion
Bob has now discovered a public IP and port which peers can use to dial him.
```

> Bob can provide this knowledge to Alice, allowing Alice to dial him.

Reviewer 3: Not entirely true, even with an IP and port, if this is behind NAT the port will need to be forwarded to the local machine's IP address.  Might be worth mentioning this quote from wikipedia
> STUN is not a self-contained NAT traversal solution applicable in all NAT deployment scenarios and does not work correctly with all of them. It is a tool among other methods and it is a tool for other protocols in dealing with NAT traversal, most notably Traversal Using Relay NAT (TURN) and Interactive Connectivity Establishment (ICE).

STUN cannot be used to traverse NAT configurations that randomly assign a port and/or IP for new connections.
For example, a symmetric NAT will randomly generate a new port and/or public IP for each outbound connection made by Alice.
This prevents Bob from determining or predicting the external IP and ports assigned to her by the NAT.

- + Does not require Bob to configure his NAT

- + Alice and Bob can communicate through a direct connection

- - Cannot traverse symmetric, cone and restricted cone type NATs

- - Requires a third peer to start a connection between Alice and Bob

- - Requires the third peer to know its public IP and ports (either through another STUN/TURN or port forwarding)

- - Requires peers to know peers that are STUN servers

- - Leads to some degree of centralisation

=== TURN

When Bob isn’t able to determine or predict his public facing IP address and port, Bob can dial out to a relay peer, which will keep a long-lived connection open.
Other peers will be able to dial through the relay peer, which will forward traffic to Bob.

> - + Can be used to traverse all commonly know NAT configurations

Reviewer 3: ```suggestion
- + Can be used to traverse all commonly known NAT configurations
```

> - + Does not require the Bob to configure his NAT

Reviewer 3: ```suggestion
- + Does not require Bob to configure his NAT
```

- - Requires a third peer to serve as a relay for communications between Alice and Bob

> - - Requires the third peer to know its inbound ports (either through another STUN/TURN or port forwarding)

Reviewer 3: ```suggestion
- - Requires the third peer to have an externally facing IP address
```

- - Requires a third peer to serve as a relay for communications between Alice and Bob

- - Requires peers to know peers that are TURN servers

- - Leads to some degree of centralisation

== Recommendation

> Assumption: A significant portion of comit nodes will have forwarded ports.

Reviewer 2: I think the opposite is going to be true unfortunately.

Author: I am guessing most people will be running a node on their home computer. Most routers have UPnP these days and I don't think people disable it.

Reviewer 2: If that is true, that would be good news! Do you by any chance have any sources for these claims? :)
The UPnP part, not about where users are going to run COMIT :D

We should implement the following NAT traversal methods in the order listed below.
After each implementation we should evaluate whether it is necessary to implement the subsequent NAT traversal methods.

1. TURN (essential for maximum NAT coverage)

> 2. STUN (optional to reduce load and reliance on TURN servers)

Reviewer 3: As suggested above, I don't think STUN is a standalone solution.

Author: STUN is a standalone solution for NAT traversal. It does not provide complete coverage of every NAT config.

> 3. UPnP (optional to reduce load and reliance on TURN servers)

Reviewer 3: We don't have to 'implement' anything to get this right, just make public the ports we use.

Author: AFAIK to take advantage of UPnP we have to implement some sort of port mapping module.
The go libp2p implementation has one. Perhaps better sense of it than I can [go-libp2p-nat](https://github.com/libp2p/go-libp2p-nat/blob/master/nat.go)

TURN is the first prioirty as it is capable of traversing most commonly known NAT configurations.

> In order for TURN to work, some port forwarded peers will have to operate as TURN servers to allow NAT restricted peers to communicate.

Reviewer 2: Can we host dedicated TURN servers?
To my understanding, those can operate solely on libp2p and don't have to know anything about COMIT. Is that correct?

Author: Yes

> In order for TURN to work, some port forwarded peers will have to operate as TURN servers to allow NAT restricted peers to communicate.

Reviewer 3: Not fully true, as commented above, the servers need an Internet facing IP address.

Reviewer 2: Isn't that what he says with "some port forwarded peers"?

Reviewer 3: The bare minimum for this to work is a publicly available IP address, how one gets it (i.e., by port forwarding) is immaterial and should not be discussed as it just adds noise (IMNSHO).

Author: good point tobin

The libp2p specification defines a circuit relay module based off the TURN protocol.
This protocol allows NAT restricted nodes to "proxy" through other p2p nodes that volunteer to act as p2p-circuit relays.

> These nodes will announce to the network that they are accepting relay connections using libp2p's peer routing.

Reviewer 2: How do we get started initially? i.e. how does a node behind NAT know about these relay nodes?

Author: The simplest solution would be to allow the user to specify a list of relay nodes through a config file or REST API call. We could provide a default list of nodes that we approve of similar to bootstrap nodes in Ethereum.

To further improve on this solution you could ask your "friends" or peers that you trust for a list of relay nodes that they trust. At this stage we are entering the realm of decentralised trust systems. The general problem becomes: "How do I know which peers are trustworthy providers of information?".  The information could be a list of reliable relay nodes or orders from an orderbook. I don't think this an easy problem to address and I don't think a potential solution to this problem should be a part of this spike. I think solving this problem will also solve the or partially solve the decentralised orderbook problem.

Currently there is no Rust implementation of the circuit relay module but the Javascript and Go implementations could be used as a guide.

STUN would be implemented if we found that there are insufficient nodes willing to act as relays due bandwidth concerns.
STUN allows peers to discover their public IP and ports and establish a communicate directly without a relay.

libp2p has it's own version of the STUN protocol using peer-routing, identify, and reuseport.

1. reuseport uses the same source port for both dialing and listening allowing us to accept new inbound connections on the source port used by the router
2. identify receives observed address information from all connected peers
3. peer-routing allows us to publish these addresses to the network and allow peers to discover their own public address on which they can accept inbound connections

> If both STUN and TURN are implemented, STUN will be the preferred method and TURN will be used as a fallback if the public address cannot be determined or predicted using STUN.

Reviewer 2: Given that COMIT is very much a 1on1 application at the moment, I think the preferred solution should be something where a node can connect to another one out of the blue. What would you propose for that?

Author: Without a TURN server or some form of third to relay data we cannot guarantee that COMIT nodes running on mobile networks or the more restrictive NAT configurations will be able to communicate. 

There are publicly available STUN servers would allow for most users to their NAT and connect to each other. [public STUN server list](https://gist.github.com/zziuni/3741933). We could provide a default list like Ethereum does with bootstrap nodes or allow the specify their own list via config or REST API.

[Cases when STUN will fail](https://stackoverflow.com/questions/23655243/nat-traversal-probability-of-success-using-stun)
